model/src/datetime/mod.rs

```diff
@@ -0,0 +1,512 @@
+//!
+//! let timestamp = Timestamp::from_secs(1_580_608_922)?;
+//!
+//! assert_eq!(
+//!     "2020-02-02T02:02:02.000000+00:00",
+//!     timestamp.iso_8601().to_string(),
+//! );
+//! # None }
+//! ```
+
+#![deny(clippy::missing_docs_in_private_items)]
+
+pub mod error;
+
+mod constant;
+mod display;
+
+use crate::datetime::constant::{MINUTES_PER_HOUR, SECONDS_PER_MINUTE};
+
+pub use self::{display::TimestampIso8601Display, error::TimestampParseError};
+
+use self::constant::{
+    Month, DAYS_PER_YEAR, EPOCH_UNIX, MICROSECONDS_PER_SECOND, SECONDS_PER_DAY, SECONDS_PER_HOUR,
+    YEAR_MAX, YEAR_MIN,
+};
+use serde::{
+    de::{Deserialize, Deserializer, Error as DeError, Visitor},
+    ser::{Serialize, Serializer},
+};
+use std::{
+    convert::TryFrom,
+    fmt::{Formatter, Result as FmtResult},
+    num::NonZeroU64,
+    str::FromStr,
+};
+
+/// Byte value of the number 0.
+const ZERO_BYTE: u8 = b'0';
+
+/// Representation of a Unix timestamp with milliseconds.
+///
+/// # Display
+///
+/// The timestamp does not itself implement [`core::fmt::Display`]. It could
+/// have two possible display implementations: that of the Unix timestamp or
+/// that of the timestamp in ISO 8601 format. Therefore, the preferred
+/// implementation may be chosen by explicitly retrieving the Unix timestamp
+/// with [seconds precision], with [microseconds precision], or
+/// [retrieving an ISO 8601 formatter].
+///
+/// [retrieving an ISO 8601 formatter]: Self::iso_8601
+/// [microseconds precision]: Self::as_micros
+/// [seconds precision]: Self::as_secs
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct Timestamp(NonZeroU64);
+
+impl Timestamp {
+    /// Create a non-zero timestamp from a Unix timestamp with microseconds
+    /// precision if the value is not zero.
+    pub const fn from_micros(unix_microseconds: u64) -> Option<Self> {
+        // `const`: Clippy suggests using `Option::map_or` which isn't const.
+        #[allow(clippy::option_if_let_else)]
+        if let Some(value) = NonZeroU64::new(unix_microseconds) {
+            Some(Self(value))
+        } else {
+            None
+        }
+    }
+
+    /// Create a non-zero timestamp from a Unix timestamp with seconds precision
+    /// if the value is not zero.
+    pub const fn from_secs(unix_seconds: u64) -> Option<Self> {
+        let micros = unix_seconds * MICROSECONDS_PER_SECOND;
+
+        // `const`: Clippy suggests using `Option::map_or` which isn't const.
+        #[allow(clippy::option_if_let_else)]
+        if let Some(value) = NonZeroU64::new(micros) {
+            Some(Self(value))
+        } else {
+            None
+        }
+    }
+
+    /// Parse a timestamp from an ISO 8601 datetime string emitted by Discord.
+    ///
+    /// Discord emits two ISO 8601 valid formats of datetimes: with microseconds
+    /// (2021-01-01T01:01:01.010000+00:00) and without microseconds
+    /// (2021-01-01T01:01:01+00:00). This supports parsing from either.
```

Comment:
Could it be noted here which years the parser supports here (and in the `FromStr` impl as well)